map2loop/thickness_calculator.py

```diff
@@ -340,3 +347,208 @@ def compute(
+            3) is the distance between the two intersections less than half the map dimensions? (avoids incorrect intersections to be picked)
+            4) is the stratigraphic measurement strike within 30 degrees of the strike of the geological contacts?
+
+        - once the intersections pass the checks, calculate the thickness of the unit at the intersection points, using the general formula L*sin(dip)
+
+        Attributes:
+        -----------
+            units (pandas.DataFrame): the data frame of units to add thicknesses to
+            stratigraphic_order (list): a list of unit names sorted from youngest to oldest
+            basal_contacts (geopandas.GeoDataFrame): basal contact geo data with locations and unit names of
+            the contacts (columns must contain ["ID","basal_unit","type","geometry"])
+            structure_data (pandas.DataFrame): sampled structural data
+            map_data (map2loop.MapData): a catchall so that access to all map data is available
+
+        Returns:
+        --------
+            pandas.DataFrame: units dataframe with added thickness columns:
+            "gammaThickness" is the median thickness of the unit,
+            "gammaStdDev" is the standard deviation of the thickness of the unit
+
+        Note:
+        -----
+            This method is highly dependent on the existence of stratigraphic measurements that follow the strike of the geological contacts.
+            If an unit does not contain a stratigraphic measurement, the thickness will not be calculated. Future versions of map2loop will attempt to solve for this.
+            If the thickness is not calculated for a given unit, the assigned thickness will be -1.
+            For the bottom and top units of the stratigraphic sequence, the assigned thickness will also be -1.
+        """
+        sampled_contacts = map_data.sampled_contacts
+        sampled_structures = structure_data
+        basal_contacts = basal_contacts.copy()
+
+        geology = map_data.get_map_data(datatype=Datatype.GEOLOGY)
+        geology[['minx', 'miny', 'maxx', 'maxy']] = geology.bounds
+
+        sampled_structures = geopandas.GeoDataFrame(
+            sampled_structures,
+            geometry=geopandas.points_from_xy(sampled_structures.X, sampled_structures.Y),
+            crs=basal_contacts.crs,
+        )
+        sampled_structures['unit_name'] = geopandas.sjoin(sampled_structures, geology)['UNITNAME']
+        sampled_basal_contacts = rebuild_sampled_basal_contacts(basal_contacts, sampled_contacts)
+
+        basal_contacts_bu = basal_contacts.copy()
+        basal_contacts_bu['geometry'] = basal_contacts_bu.buffer(0.1)
+
+        map_dx = geology.total_bounds[2] - geology.total_bounds[0]
+        map_dy = geology.total_bounds[3] - geology.total_bounds[1]
+
+        ths = []
+        lis = []
+
+        for s in range(0, len(sampled_structures)):
```

Comment:
It looks like this could be rewritten to remove the for loop and to be vectorised. If we force geopandas 0.14+ it uses the pygeos backend. So geometry operations can be done as vectorised code. 

Without rewriting the function I would start by changing calculate endpoints to be a vectorised function so that it returns an array of lines. Those lines can then be intersected with the basal contacts.